### PR TITLE
Simplifications

### PR DESCRIPTION
**Pull request by zolkis**
As discussed on the Scripting call of 3. Sep.,
- Remove ThingInstance, IOW merge it to ConsumedThing and ExposedThing
- Add algorithms for ExposedThing.expose() and destroy().

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -462,13 +468,36 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
...
+            Create an <a>ExposedThing</a> object <var>thing</var> initialized from <var>model</var>.
+          </li>
+          <li>
+            For each <a>ThingFragment</a> property, initialize the property based on the provided initial or default values provided to the local <a>WoT Runtime</a> implementation, for instance initialize:
+            <ol>
+              <li>
+                 the <code>id</code> property to be the final unique identifier of the <a>Thing</a>,
+              </li>
+              <li>
+                the <code>security</code> object to represent the actual security scheme andits properties as set up by the implementation,
```
Comment (danielpeintner) [inline index.html]: Typo: change "andits" to "and its"

**Comment by zolkis** [comment]
I would also like to simplify value type definitions, as indicated in #142

**Comment by zolkis** [comment]
Now SecurityScheme, DataSchema (and DataType) are "outsource" to the TD spec, and the Scripting spec contains only the runtime related aspects (such as value matching algorithm, instantiation of properties of type DataSchema and SecurityScheme).

File: index.html [hunk 2]
```diff
@@ -378,7 +378,7 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
         </ol>
       </p>
     </section> <!-- Discovery -->
-
+<
     <section> <h3> The <dfn>fetch()</dfn> method</h3>
       <p>
         Accepts an <code>url</code> argument of type <code>USVString</code> that represents a URL (e.g. <code>"file://..."</code> or <code>"https://..."</code>) and returns a <code><a>Promise</a></code> that resolves with a <a>ThingDescription</a> (a serialized <a>JSON-LD</a> document of type <code>USVString</code>).
```
Comment (mkovatsc) [inline index.html]: Probably should talk about the `application/td+json` media type.
Could have a note that future media types can identify other TD serializations (e.g., CBOR).

File: index.html [hunk 3]
```diff
@@ -425,10 +425,16 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
...
+          <li>
+            Create a <a>ConsumedThing</a> object <var>thing</var> initialized from <var>model</var>.
           </li>
           <li>
-            Create a <a>ConsumedThing</a> object <var>thing</var> initialized from <var>instance</var>.
+            Add the <code>read()</code> and <code>write()</code> methods to the <a>ThingProperty</a> elements so that they make requests to access the remote <a>Thing</a>s and wait for the reply, as defined by the <a>Protocol Bindings</a>.
+          </li>
+          <li>
+            Add the <code>invoke()</code> methods to the <a>ThingAction</a> elements so that they make requests to the remote <a>Thing</a> to invoke its actions, as defined by the <a>Protocol Bindings</a>.
           </li>
```
Comment (mkovatsc) [inline index.html]: Need another item for ThingEvent and subscribe()

File: index.html [hunk 4]
```diff
@@ -425,10 +425,16 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
             Let <var>model</var> be the result of running the <a>TD parsing algorithm</a> with <var>td</var> as argument. If that throws an error, re-throw the error and terminate these steps.
```
Comment (mkovatsc) [inline index.html]: "model" is later used for either TD string or ThingFragment object.
Thus, I would call this here "fragment", as it is an object. The "TD parsing algorithm" could already take care of the default values (see also below).

File: index.html [hunk 5]
```diff
@@ -462,13 +468,36 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
             If <var>model</var> is not an object, throw <code>TypeError</code> and terminate these steps.
```
Comment (mkovatsc) [inline index.html]: line 448/454: an ExposedThing produced from TD also offers addProperty() etc., so "adding" can also apply here. Conversely, a ThingFragment can also contain Interaction definitions.

Comment (mkovatsc) [inline index.html]: line 453/459: produce() does not take a URL. Argument should be "model".

File: index.html [hunk 6]
```diff
@@ -462,13 +468,36 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
...
           </li>
           <li>
-            Let <var>instance</var> be the result of running the <a>Thing instantiation algorithm</a> with <var>model passed as parameter.
+            If <var>model</var> does not have an own property that is defined in <a>ThingFragment</a> with a default value, add that property and value to <var>model</var>.
```
Comment (mkovatsc) [inline index.html]: Check and align with "TD parsing algorithm" definition (see comment above).

File: index.html [hunk 7]
```diff
@@ -425,10 +425,16 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
             Let <var>model</var> be the result of running the <a>TD parsing algorithm</a> with <var>td</var> as argument. If that throws an error, re-throw the error and terminate these steps.
           </li>
           <li>
-            Let <var>instance</var> be the result of running the <a>Thing instantiation algorithm</a> with <var>model</var> as argument.
+            If <var>model</var> does not have an own property that is defined in <a>ThingFragment</a> with a default value, add that property and value to <var>model</var>.
```
[the remark this answers is not in the document]
Comment (mkovatsc) [reply index.html]: I now see that procude() can take an object that would need the same curation step.
Maybe the "TD parsing algorithm" should be split into JSON.parse() of the string and then a validation/curation algorithm that can be reused in both consume() and produce().

File: index.html [hunk 8]
```diff
@@ -462,13 +468,36 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
...
+            For each <a>ThingFragment</a> property, initialize the property based on the provided initial or default values provided to the local <a>WoT Runtime</a> implementation, for instance initialize:
+            <ol>
+              <li>
+                 the <code>id</code> property to be the final unique identifier of the <a>Thing</a>,
+              </li>
+              <li>
+                the <code>security</code> object of type <a>SecurityScheme</a> to represent the actual security scheme and its properties as set up by the implementation,
+              </li>
+              <li>
+                the <code>properties</code> property to be an array of <a>ThingProperty</a> objects in which the <code>read()</code> and <code>write()</code> methods are provided to define local methods to get and set the <a>Property</a> values,
```
Comment (mkovatsc) [inline index.html]: "**array** of <a>ThingProperty</a> objects"?
In the TD model, it is a map and most implementations would probably use a map for fast lookup. Yet this is an implementation detail.
I think we can skip this part and talk about adding the read/write methods similar to ConsumendThing, only that the access is local.
Initializing the state that holds the Property values should maybe go before adding the read and write?

File: index.html [hunk 9]
```diff
@@ -703,12 +712,52 @@ <h2>The <dfn>ExposedThing</dfn> interface</h2>
...
+        The <code>expose()</code> method MUST run the following steps:
+        <ol>
+          <li>
+            Return a <code>Promise</code> <var>promise</var> and execute the next steps <a>in parallel</a>.
+          </li>
+          <li>
+            If invoking <code>expose()</code> is not allowed for the current scripting context for security reasons, reject <var>promise</var> with <code>SecurityError</code> and terminate these steps.
+          </li>
+          <li>
+            Make a request to the underlying platform to create and configure serving external requests for WoT interactions, based on the <a>Protocol Bindings</a>.
```
Comment (mkovatsc) [inline index.html]: "create and configure serving external requests" -- in particular "create" sounds confusing.
maybe: """
Make a request to the underlying platform to attach protocol listeners to the WoT interactions, based on the <a>Protocol Bindings</a>.
"""

File: index.html [hunk 10]
```diff
@@ -1611,175 +1649,31 @@ <h3>The <dfn data-dfn-for="">Observable</dfn> interface</h3>
...
+        <li>
+          Minimize scripts functionality and complexity.
+        </li>
+        <li>
+          Avoid heavy functional processing in scripts without authentication.
+        </li>
+      </ol>
+    </p>
+    <p>
+      Security metadata is represented as dictionary objects of type <dfn>SecurityScheme</dfn> whose properties and sub-classes are defined in the <a href="https://w3c.github.io/wot-thing-description/#sec-security-vocabulary-definition">Thing Description specification</a>.
```
Comment (mkovatsc) [inline index.html]: This is quite lost here now. I recommend to move it to "Data types used in a Thing" like the other structures.

File: index.html [hunk 11]
```diff
@@ -378,7 +378,7 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
         </ol>
       </p>
     </section> <!-- Discovery -->
-
+<
```
[the remark this answers is not in the document]
Comment (mkovatsc) [reply index.html]: Was actually a problem with https://zolkis.github.io/wot-scripting-api/
But still does not belong there...

**Comment by zolkis** [comment]
Thanks @mkovatsc 
I will address these comments, but not before the WG call. 
Since the changes are not done, this will postpone publication of Scripting API by one week.

**Comment by danielpeintner** [comment]
Besides updating some links like "[[!WOT_TD]]" which could simply point to "#bib-wot-td" at the moment it looks good to me.

One more issue that come up recently is that besides the existing types such as "BooleanSchema" or "NumberSchema" we might also have "NullSchema".
see 
https://github.com/w3c/wot-thing-description/issues/222#issuecomment-418872522

**Comment by zolkis** [comment]
Thanks Daniel.

> Besides updating some links like "[[!WOT_TD]]" which could simply point to "#bib-wot-td" at the moment it looks good to me.

I thought using explicit biblio references over simple links was encouraged in specs. If we use links, it could be a direct link, IMHO it's more convenient.

> we might also have "NullSchema".

Currently it has no reference in the TD doc, so I won't include it yet. As soon as it's published, we can link it.

**Comment by zolkis** [comment]
Ah, I noticed the WOT_TD (with underscore), sorry for missing that. I have fixed/amended that. Can we merge?

**Comment by zolkis** [comment]
Merging for easier review, and based on the [conditional lgtm](https://github.com/w3c/wot-scripting-api/pull/138#issuecomment-419002356).

**Pull request merged**